### PR TITLE
x2APIC support for ACRN guests

### PR DESCRIPTION
**Pull request by sainath14**
This patch series adds x2APIC support for ACRN guests. Also adds support for LAPIC pass-thru for guests running using partition mode of ACRN.

Tracked-On: #1626
Signed-off-by: Sainath Grandhi <sainath.grandhi@intel.com>
Reviewed-by: Xu Anthony <anthony.xu@intel.com>

### CODE REVIEW

File: hypervisor/arch/x86/guest/vlapic.c [hunk 1]
```diff
@@ -1936,6 +1936,32 @@ static void vlapic_timer_expired(void *data)
...
+	}
+}
+
+static int vlapic_x2apic_access(struct vcpu *vcpu)
+{
+	struct acrn_vlapic *vlapic;
+	int error = 0;
+
+	vlapic = vcpu_vlapic(vcpu);
+	if (is_x2apic_enabled(vlapic) == false) {
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vlapic.c]: if (!is_x2apic_enabled())   ?

File: hypervisor/arch/x86/guest/vmsr.c [hunk 2]
```diff
@@ -51,6 +51,43 @@ static void enable_msr_interception(uint8_t *bitmap, uint32_t msr_arg)
...
+/*
+ * Enable read and write msr interception for x2APIC MSRs
+ * MSRs that are not supported in the x2APIC range of MSRs,
+ * i.e. anything other than the ones below and between
+ * 0x802 and 0x83F, are not intercepted
+ */
+
+static void intercept_x2apic_msrs(uint8_t *msr_bitmap_arg)
+{
+	uint8_t *msr_bitmap = msr_bitmap_arg;
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vmsr.c]: Please use a static array to contain all the MSRs we want to intercept, and simplify this API to loop the array.

BTW, can we add those MSRs to emulated_msrs?

File: hypervisor/arch/x86/guest/vlapic.c [hunk 3]
```diff
@@ -1955,8 +1981,13 @@ vlapic_rdmsr(struct vcpu *vcpu, uint32_t msr, uint64_t *rval)
 		break;
 
 	default:
-		dev_dbg(ACRN_DBG_LAPIC,
-			"Invalid vlapic msr 0x%x access\n", msr);
+		if (is_x2apic_msr(msr)) {
+			error = vlapic_x2apic_access(vcpu);
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/vlapic.c]: I may lose something. This API is not emulating the rdmsr..  Where is it?

Comment (sainath14) [reply hypervisor/arch/x86/guest/vlapic.c]: read or write is handled in vlapic_x2apic_access

File: hypervisor/arch/x86/cpuid.c [hunk 4]
```diff
@@ -369,13 +360,33 @@ void guest_cpuid(struct vcpu *vcpu,
...
+				} else {
+					*eax = fls32(vcpu->vm->hw.created_vcpus - 1U) + 1U;
+				}
+				*ebx = vcpu->vm->hw.created_vcpus;
+				*ecx |= (2U << 8U);
+			break;
+			default:
+				*eax = 0U;
+				*ebx = 0U;
+				*ecx |= (0U << 8U);
```
[the remark this answers is not in the document]
Comment (sainath14) [reply hypervisor/arch/x86/cpuid.c]: first 8 bits of ecx should return what is sent in ecx and edx should always return x2apic id even though the value in subleaf is invalid.

File: hypervisor/arch/x86/guest/vmsr.c (hunk 2 above)
Comment (sainath14) [reply hypervisor/arch/x86/guest/vmsr.c]: addressed by adding a static array of x2apic MSRs

File: hypervisor/arch/x86/guest/vlapic.c [hunk 5]
```diff
@@ -49,6 +49,8 @@ static inline uint32_t prio(uint32_t x)
 #define	APICBASE_BSP		0x00000100UL
 #define	APICBASE_X2APIC		0x00000400U
 #define	APICBASE_ENABLED	0x00000800UL
+#define LOGICAL_ID_MASK         0xFU
```
[the remark this answers is not in the document]
Comment (sainath14) [reply hypervisor/arch/x86/guest/vlapic.c]: addressed

File: hypervisor/arch/x86/guest/vlapic.c [hunk 6]
```diff
@@ -1032,35 +1019,59 @@ vlapic_calcdest(struct vm *vm, uint64_t *dmask, uint32_t dest,
...
+					 */
+					mda_cluster_id = (dest >> 4) & 0xfU;
+					mda_cluster_ldest = dest & 0xfU;
+
+					dfr = vlapic->apic_page.dfr.v;
+					ldr = vlapic->apic_page.ldr.v;
 
+					if ((dfr & APIC_DFR_MODEL_MASK) ==
+							APIC_DFR_MODEL_FLAT) {
+						ldest = ldr >> 24;
```
[the remark this answers is not in the document]
Comment (sainath14) [reply hypervisor/arch/x86/guest/vlapic.c]: changed

File: hypervisor/arch/x86/guest/vlapic.c [hunk 7]
```diff
@@ -1996,18 +1996,33 @@ static inline bool is_x2apic_enabled(const struct acrn_vlapic *vlapic)
 	}
 }
 
-static int vlapic_x2apic_access(struct vcpu *vcpu)
+static inline  uint32_t x2apic_msr_to_regoff(uint32_t msr)
+{
+	return (msr - 0x800U) << 4U;
```
[the remark this answers is not in the document]
Comment (sainath14) [reply hypervisor/arch/x86/guest/vlapic.c]: added an ASSERT on the range of msr value

File: hypervisor/arch/x86/guest/vlapic.c (hunk 1 above)
Comment (sainath14) [reply hypervisor/arch/x86/guest/vlapic.c]: Remember we need to use == false, to follow some MISRA-C rule

**Comment by sainath14** [comment]
@HuiHuangShi  Addressed most of the violations. Rest of them, I guess, they are false postives

**Comment by HuiHuangShi** [comment]
1698<br>14 new potential violations to the coding guideline detected. Please help check. Thanks!  <br><br>1.filename:/hypervisor/include/arch/x86/msr.h&nbsp;&nbsp;function:is_x2apic_msr&nbsp;&nbsp;offset:533:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : is_x2apic_msr<br><br>2.filename:/hypervisor/include/arch/x86/msr.h&nbsp;&nbsp;function:is_x2apic_read_only_msr&nbsp;&nbsp;offset:545:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : is_x2apic_read_only_msr<br><br>3.filename:/hypervisor/include/arch/x86/msr.h&nbsp;&nbsp;function:is_x2apic_write_only_msr&nbsp;&nbsp;offset:561:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : is_x2apic_write_only_msr<br><br>4.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:105:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of bit operator on signed type. : & used with enum rw_mode: mode<br><br>5.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:105:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of underlying enum representation value. : & used with enum rw_mode: mode<br><br>6.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:105:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of bit operator on signed type. : & used with enum rw_mode: READ<br><br>7.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:105:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of underlying enum representation value. : & used with enum rw_mode: READ<br><br>8.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:111:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of underlying enum representation value. : & used with enum rw_mode: mode<br><br>9.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:111:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of underlying enum representation value. : & used with enum rw_mode: WRITE<br><br>10.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:111:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of bit operator on signed type. : & used with enum rw_mode: WRITE<br><br>11.filename:/hypervisor/arch/x86/guest/vmsr.c&nbsp;&nbsp;function:enable_msr_interception&nbsp;&nbsp;offset:111:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Use of bit operator on signed type. : & used with enum rw_mode: mode<br><br>12.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_process_init_sipi&nbsp;&nbsp;offset:1172:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Type conversion without cast. :  (enum vm_state and enum vcpu_state): target_vcpu -> state != VCPU_INIT<br><br>13.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_set_apicbase&nbsp;&nbsp;offset:1733:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Implicit conversion: actual to formal param (MR). :  (void and struct vcpu):  param  1<br><br>14.filename:/hypervisor/arch/x86/guest/vlapic.c&nbsp;&nbsp;function:vlapic_x2apic_self_ipi_handler&nbsp;&nbsp;offset:2457:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:(void) missing for discarded return value.<br><br>

**Comment by HuiHuangShi** [comment]
> 1698
> 14 new potential violations to the coding guideline detected. Please help check. Thanks!
> 
> 1.filename:/hypervisor/include/arch/x86/msr.h  function:is_x2apic_msr  offset:533:
>        reason:Static procedure is not explicitly called in code analysed. : is_x2apic_msr
> 
> 2.filename:/hypervisor/include/arch/x86/msr.h  function:is_x2apic_read_only_msr  offset:545:
>        reason:Static procedure is not explicitly called in code analysed. : is_x2apic_read_only_msr
> 
> 3.filename:/hypervisor/include/arch/x86/msr.h  function:is_x2apic_write_only_msr  offset:561:
>        reason:Static procedure is not explicitly called in code analysed. : is_x2apic_write_only_msr
> 
> 4.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:105:
>        reason:Use of bit operator on signed type. : & used with enum rw_mode: mode
> 
> 5.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:105:
>        reason:Use of underlying enum representation value. : & used with enum rw_mode: mode
> 
> 6.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:105:
>        reason:Use of bit operator on signed type. : & used with enum rw_mode: READ
> 
> 7.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:105:
>        reason:Use of underlying enum representation value. : & used with enum rw_mode: READ
> 
> 8.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:111:
>        reason:Use of underlying enum representation value. : & used with enum rw_mode: mode
> 
> 9.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:111:
>        reason:Use of underlying enum representation value. : & used with enum rw_mode: WRITE
> 
> 10.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:111:
>        reason:Use of bit operator on signed type. : & used with enum rw_mode: WRITE
> 
> 11.filename:/hypervisor/arch/x86/guest/vmsr.c  function:enable_msr_interception  offset:111:
>        reason:Use of bit operator on signed type. : & used with enum rw_mode: mode
> 
> 12.filename:/hypervisor/arch/x86/guest/vlapic.c  function:vlapic_process_init_sipi  offset:1172:
>        reason:Type conversion without cast. : (enum vm_state and enum vcpu_state): target_vcpu -> state != VCPU_INIT
> 
> 13.filename:/hypervisor/arch/x86/guest/vlapic.c  function:vlapic_set_apicbase  offset:1733:
>        reason:Implicit conversion: actual to formal param (MR). : (void and struct vcpu): param 1
> 
> 14.filename:/hypervisor/arch/x86/guest/vlapic.c  function:vlapic_x2apic_self_ipi_handler  offset:2457:
>        reason:(void) missing for discarded return value.

All of this are false positive

**Comment by sainath14** [comment]
@HuiHuangShi Thanks for the clarification.

**Pull request merged**